python/singa/sonnx.py

```diff
@@ -0,0 +1,299 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+
+
+from __future__ import division
+import pickle
+from singa import tensor
+from singa.tensor import Tensor
+from singa import autograd
+from singa import optimizer
+import onnx
+from onnx import helper
+from onnx import AttributeProto, TensorProto, GraphProto
+from onnx import numpy_helper
+from collections import Counter, deque
+import math
+
+from .tensor import Tensor
+from . import layer
+from singa.proto import model_pb2
+from . import singa_wrap as singa
+#from .tensor import einsum
+from autograd import *
+from singa.tensor import to_numpy
+from autograd import _Conv2d,_Pooling2d,_BatchNorm2d
+
+
+class ONNXm(Layer):
```

Comment:
I think this class is performing both Importing ONNX models (backend) and Exporting to ONNX (frontend). It will be better to [separate the importing (backend) from the exporting (frontend)](https://github.com/apache/incubator-singa/pull/416#discussion_r236562502). 

Now the frontend function is a static method INSIDE the ONNXm class (see from_onnx_model() ), while the backend function is a function OUTSIDE the ONNXm class (see to_onnx_model() ).

This is not a problem, but the class which supports the backend operations must inherit from from `onnx.backend.base.Backend`. So there are two options: 

1. make ONNXm inherits from onnx.backend.base.Backend:

```
from onnx.backend.base import Backend
class ONNXm(backend)
...
```
in this case the ONNXm class will do both backend and frontend together, so the backend functions must be implemented INSIDE it.

or (better)

2. create another new backend class which calls this one as needed. The new backend class must also inherit from `onnx.backend.base.Backend`:

```
from onnx.backend.base import Backend
class SingaBackend(backend)
...
```

See [CNTK backend](https://github.com/onnx/onnx-cntk/blob/765f51f5b26f62933bb2e29c29661a152c001b89/onnx_cntk/backend.py#L13) and [TensforFlow backend](https://github.com/onnx/onnx-tensorflow/blob/c4a75b09e85ffb7b5afda14b64194ca972f957bf/onnx_tf/backend.py#L32) .

Reply:
I separated frontend and backend. backend include one-time backend and multiple times backend.
should I keep cross entropy layer?

Reply:
I separated frontend and backend. backend include one-time backend and multiple times backend.
I just removed cross entropy layer

Reply:
According to [onnx backend implementation](https://github.com/onnx/onnx/blob/master/docs/ImplementingAnOnnxBackend.md), we need to implement two backends.
one for one time inference; one for running the onnx model times (but just do the conversion for only one time).